dialogue-clients/src/test/java/com/palantir/dialogue/clients/ProtocolVersionFilteringDialogueDnsResolverTest.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * (c) Copyright 2024 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.dialogue.clients;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+import com.google.common.collect.ImmutableSet;
+import com.palantir.dialogue.core.DialogueDnsResolver;
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+import org.junit.jupiter.api.Test;
+
+class ProtocolVersionFilteringDialogueDnsResolverTest {
+
+    @Test
+    void returns_only_ipv4_when_input_contains_only_ipv4() throws UnknownHostException {
+        ImmutableSet<InetAddress> resolvedAddresses = ImmutableSet.of(
+                InetAddress.getByAddress(new byte[] {1, 1, 1, 1}), InetAddress.getByAddress(new byte[] {2, 2, 2, 2}));
+
+        DialogueDnsResolver delegate = mock(DialogueDnsResolver.class);
```

Comment:
It might be easier if we split the `MapBasedDnsResolver` out of `DialogueClientsDnsIntegrationTest` to reuse in this test. I find mockito a giant pain to work with in favor of simpler stubs. This way we can mutate the underlying multimap, and assert the results are what we expect, with the compile-time guarantee that all methods are implemented.

Reply:
let's do that in a separate change, then i'll rebase: https://github.com/palantir/dialogue/pull/2158